File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zero_count = 0
        product = 1
        for n in nums:
            if n == 0:
                zero_count += 1
            else:
                product *= n
        if zero_count > 1:
            return [0 for i in range(len(nums))]
        elif zero_count == 1:
            res = []
            for n in nums:
                if n == 0:
                    res.append(product)
                else:
                    res.append(0)
            return res
        else:
            res = []
            for n in nums:
                res.append(product // n)
            return res

        

        

File: Data Structures & Algorithms/permutations/submission-0.java
class Solution {
    // global data
    List<List<Integer>> res;


    private List<List<Integer>> backtrack(List<Integer> perm, int[] nums, boolean[] pick) {
        // if reject(P, c) then return
        if (perm.size() == nums.length) {
            res.add(new ArrayList<>(perm));
            return res;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!pick[i]) {
                perm.add(nums[i]);
                pick[i] = true;
                backtrack(perm, nums, pick);
                perm.remove(perm.size() - 1);
                pick[i] = false;
            }
        }
        return res;
    }

    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        backtrack(new ArrayList<>(), nums, new boolean[nums.length]);
        return res;
    }
}
